src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -563,27 +537,252 @@ internal SafeFileHandle OpenPort(string portName)
+                Interop.Sys.PollEvents events = PollEvents(1,
+                                                           pollReadEvents: _readQueue.Count > 0,
+                                                           pollWriteEvents: _writeQueue.Count > 0);
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLNVAL) ||
+                    events.HasFlag(Interop.Sys.PollEvents.POLLERR))
+                {
+                    // bad descriptor or some other error we can't handle
+                    FinishPendingIORequests();
+                    break;
+                }
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLIN))
+                {
+                    int bytesRead = DoIORequest(_readQueue, _processReadDelegate);
+                    totalBytesRead += bytesRead;
+                }
+
+                // check if there is any new data (either already read or in the driver input)
+                // this event is private and handled inside of SerialPort
+                // which then throttles it with the threshold
+                long totalBytesAvailable = totalBytesRead + BytesToRead;
+                if (totalBytesAvailable > lastTotalBytesAvailable)
+                {
+                    lastTotalBytesAvailable = totalBytesAvailable;
+
+                    // We need new task so that this thread doesn't get deadlocked when someone calls
+                    // Read from within the event
+
+                    ThreadPool.QueueUserWorkItem(s => {
+                            var thisRef = (SerialStream)s;
+                            thisRef.DataReceived(thisRef, new SerialDataReceivedEventArgs(SerialData.Chars));
+                        }, this);
+                }
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLOUT))
+                {
+                    DoIORequest(_writeQueue, _processWriteDelegate);
+                }
+            }
+        }
+
+        private static CancellationTokenSource GetCancellationTokenSourceFromTimeout(int timeoutMs)
+        {
+            return timeoutMs == SerialPort.InfiniteTimeout ?
+                null :
+                new CancellationTokenSource(Math.Max(timeoutMs, TimeoutResolution));
+        }
+
+        private static Exception GetLastIOError()
+        {
+            return Interop.GetIOException(Interop.Sys.GetLastErrorInfo());
+        }
+
+        private class SerialStreamIORequest : TaskCompletionSource<int>
+        {
+            public Memory<byte> Buffer { get; private set; }
+            public bool IsCompleted => Task.IsCompleted;
+            private CancellationToken _cancellationToken;
+
+            public SerialStreamIORequest(CancellationToken ct, Memory<byte> buffer)
+                : base(TaskCreationOptions.RunContinuationsAsynchronously)
+            {
+                _cancellationToken = ct;
+                ct.Register(s => ((TaskCompletionSource<int>)s).TrySetCanceled(), this);
```

Comment:
You need to Dispose of the CancellationTokenRegistration this returns in the various Complete methods.  Otherwise, this registration will be leaked into the cancellation token.

Reply:
I noticed I actually only need it with Read/Write so just disposing it there directly